src/problems/sudoku.clj

```diff
@@ -0,0 +1,124 @@
+(ns problems.sudoku)
+;;OH BOY KIDS HERE WE GO
+(defn distinct?
+  "Returns true if no two of the arguments are ="
+  [& more]
+  (let [old (count more)
+        set (into #{} more)
+        new (count set)]
+    (= new old)))
+
+;;## validation top level ##
+(defn every-element-in-vec-distinct?
+ "Returns True if every element in every vector is unique or nil.
+  Returns False otherwise."
+  [vecs]
+ (every? identity (map #(apply distinct? %) vecs)))
+
+(def every-vector-valid? (memoize every-element-in-vec-distinct?))
+
+;;## validating 3x3 blocks ##
+(defn hop [adjusted-n k chunk]
+  "Given the first number in the block (n), and the offset (k), returns the 3
+  numbers that make up a column in a sudoku block."
+  (take-nth 9 (drop (+ adjusted-n k) chunk)))
+
+(defn hop-block [chunk n]
+  "Gets the three columns in a 3x3 sudoku block."
+  (let [adjusted-n (* n 3)] ;; the upper left hand number in a sudoku block.
+    (into [] (concat (hop adjusted-n 0 chunk)
+                     (hop adjusted-n 1 chunk)
+                     (hop adjusted-n 2 chunk)))))
+
+(defn get-blocks [chunk]
+  "Operates on a 3x9 chunk of a sudoku state. Returns a list of 3 vectors of
+  9 elements that represents a single block (3x3) in sudoku."
+  (map #(hop-block chunk %) [0 1 2])) ;; the 'numbers' of the three blocks
+
+(defn check-squares? [state]
+  "Checks that every 3x3 sudoku block has contains distinct elements or nil."
+  (every-vector-valid? (mapcat get-blocks (partition 27 state))))
+
+;;validate horizontal
+(defn check-horizontal?
+ "Check that each row contains distinct elements"
+ [state]
+ (every-vector-valid? (partition 9 state)))
+
+(defn do-children
+  "Given a sudoku state, a vector of ints, returns all valid child states as a vector of ints"
```

Comment:
This may be the closest thing you have to any sort of documentation of what your state is, and it's way down here under all these `check` functions. And, weirdly, it's _above_ the `check-vertical?` function(s), which arguably should be _above_ this.